server/src/api/requests/quantity.ts

```diff
@@ -0,0 +1,119 @@
+}
+
+export class Quantity {
+    public static async inStock(itemId: number = -1) {
+        const quantities: ItemQuantities = await Quantity.getQuantities(["SUBMITTED", "APPROVED", "READY_FOR_PICKUP", "FULFILLED", "LOST", "DAMAGED"], itemId);
+        const totalAvailable: ItemQtyAvailable = await Item.getTotalAvailable(itemId);
+
+        return this.getTotalAvailableLessStatuses(quantities, totalAvailable, ["FULFILLED", "LOST", "DAMAGED"], itemId);
+    }
+
+    public static async unreserved(itemId: number = -1) {
+        const quantities: ItemQuantities = await Quantity.getQuantities(["SUBMITTED", "APPROVED", "READY_FOR_PICKUP", "FULFILLED", "LOST", "DAMAGED"], itemId);
+        const totalAvailable: ItemQtyAvailable = await Item.getTotalAvailable(itemId);
+
+        return this.getTotalAvailableLessStatuses(quantities, totalAvailable, ["SUBMITTED", "APPROVED", "READY_FOR_PICKUP", "FULFILLED", "LOST", "DAMAGED"]);
+    }
+
+    public static async availableForApproval(itemId: number = -1) {
+        const quantities: ItemQuantities = await Quantity.getQuantities(["SUBMITTED", "APPROVED", "READY_FOR_PICKUP", "FULFILLED", "LOST", "DAMAGED"], itemId);
+        const totalAvailable: ItemQtyAvailable = await Item.getTotalAvailable(itemId);
+
+        return this.getTotalAvailableLessStatuses(quantities, totalAvailable, ["APPROVED", "READY_FOR_PICKUP", "FULFILLED", "LOST", "DAMAGED"]);
+    }
+
+    public static async all(itemId: number = -1) {
+        const quantities: ItemQuantities = await Quantity.getQuantities(["SUBMITTED", "APPROVED", "READY_FOR_PICKUP", "FULFILLED", "LOST", "DAMAGED"], itemId);
+        const totalAvailable: ItemQtyAvailable = await Item.getTotalAvailable(itemId);
+
+        const qtyInStock: ItemQtyAvailable = this.getTotalAvailableLessStatuses(quantities, totalAvailable, ["FULFILLED", "LOST", "DAMAGED"], itemId);
+        const qtyUnreserved: ItemQtyAvailable = this.getTotalAvailableLessStatuses(quantities, totalAvailable, ["SUBMITTED", "APPROVED", "READY_FOR_PICKUP", "FULFILLED", "LOST", "DAMAGED"]);
+        const qtyAvailableForApproval: ItemQtyAvailable = this.getTotalAvailableLessStatuses(quantities, totalAvailable, ["APPROVED", "READY_FOR_PICKUP", "FULFILLED", "LOST", "DAMAGED"]);
+
+        return {
+            qtyInStock,
+            qtyUnreserved,
+            qtyAvailableForApproval
+        };
+    }
+
+    private static getTotalAvailableLessStatuses(quantities: ItemQuantities, totalAvailable: ItemQtyAvailable, statuses: RequestStatus[] = [], itemId: number = -1): ItemQtyAvailable {
+        const result = {};
+
+        for (const id in totalAvailable) {
+            if (totalAvailable.hasOwnProperty(id)) {
+                if (quantities.hasOwnProperty(id)) {
+                    const item = quantities[id];
+                    let quantity: number = totalAvailable[id];
+                    for (let i = 0; i < statuses.length; i++) {
+                        quantity -= item[statuses[i]];
```

Comment:
just to clarify, do items or requests own status?

Reply:
Requests but the naming might be confusing here.